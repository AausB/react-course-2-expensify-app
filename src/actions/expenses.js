import uuid from 'uuid';
import moment from 'moment';

import database from '../firebase/firebase';

// How SYNC action generators work:
//   component calls action generator
//   action generator returns object !!!
//   component dispatches object
//   redux store changes 

// How ASYNC action generators work:
//   component calls action generator
//   action generator returns function !!! 
//     -> redux thunk: middleware allows you to write action creators that return a function instead of an action.
//   component dispatches function (redux by default does not allow to dispatch function)
//   function runs (has the ability to dispatch other actions and do whatever it wants) 

// ADD_EXPENSE
// prior to redux-thunk:
//
// export const addExpense = (
//   {
//     description = '',
//     note = '',
//     amount = 0,
//     createdAt = moment().valueOf() // alex set now as default
//   } = {}
// ) => ({
//   type: 'ADD_EXPENSE',
//   expense: {
//     id: uuid(),
//     description,
//     note,
//     amount,
//     createdAt
//   }
// });

// ADD_EXPENSE
// with redux-thunk
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

// startAddExpense is the function that first writes to database and then to the store
//   returns a function
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    
    // initialize data
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = {description, note, amount, createdAt};

    // add expense to database
    // return database... returns the promise chain to chain it with "then" in expenses.test.js
    return database.ref('expenses')
      .push(expense)
      .then((ref) => {
        // add expense to store
        dispatch(addExpense({
          id: ref.key, // id comes from firebase now
          ...expense  
        }));
      })
      .catch((error) => {
        // to be done
      });

  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// SET_EXPENSES
// set the expenses initially
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// start SET_EXPENSES
// async call to database to fetch all expenses
export const startSetExpenses = () => {
  return (dispatch) => {
  // 1. Fetch all expense data once -> see example in firebase.js
    // return database.ref to use .then() in app.js
    return database.ref('expenses')
      .once('value')
      .then((snapshot) => {
        const expenses = [];

  // 2. Parse data into an array
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
  // 3. Dispatch SET_EXPENSES
        dispatch(setExpenses(expenses));
      })
      .catch((error) => {
        console.log('db fetch failed:', error)
      });
  };
};
