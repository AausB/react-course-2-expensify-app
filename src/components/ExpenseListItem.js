import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

// load a locale
numeral.register('locale', 'de', {
  delimiters: {
      thousands: '.',
      decimal: ','
  },
  abbreviations: {
      thousand: 'tsd',
      million: 'm',
      billion: 'b',
      trillion: 't'
  },
  currency: {
      symbol: 'â‚¬'
  }
});

// switch between locales
numeral.locale('de');

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      {numeral(amount / 100).format('$0,0.00')} - {moment(createdAt).format('DD.MM.YYYY, HH:mm:ss')}</p>
  </div>
);

export default ExpenseListItem;
