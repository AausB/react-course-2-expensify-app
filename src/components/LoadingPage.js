import React from 'react';

const LoadingPage = (props) => (
  <div className="loader">
    <img className="loader__image" src="/images/loader.gif" alt="loading..."/>
  </div>
);

export default LoadingPage;