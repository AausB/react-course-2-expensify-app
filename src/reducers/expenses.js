// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        };
      });
    case 'SET_EXPENSES':
      return action.expenses; // do not care what was inside the state.expenses before just return the new expenses
    default:
      return state;
  }
};
