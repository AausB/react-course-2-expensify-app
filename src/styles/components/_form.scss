.form {
  display: flex;
  flex-direction: column;

  // all direct children
  > * {
    margin-bottom: $m-size;
  }

}

.form__error {
  background:$light-red;
  color: $dark-red;
  padding: $s-size;
  margin: 0 0 $m-size 0;
  // font-style: italic;
}