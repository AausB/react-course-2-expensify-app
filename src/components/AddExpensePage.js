import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// named export for testing without connect()
export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense); // to be tested
    this.props.history.push('/'); // to be tested
  };

  // to be tested
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

// for the testing:
// use mapDispatchToProps by connect()
// you can use the dispatch funtion as a property
// abstract the dispathc function away from the component itself
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
