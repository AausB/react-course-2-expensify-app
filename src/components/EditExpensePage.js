import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  // to be tested: testCase
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense); // to be tested: use spy and toHaveBeenLastCalledWith
    this.props.history.push('/'); // to be tested: use spy and toHaveBeenLastCalledWith
  };

  // to be tested: testCase
  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id }); // to be tested: use spy and toHaveBeenLastCalledWith
    this.props.history.push('/');  // to be tested: use spy and toHaveBeenLastCalledWith
  };

  // to be tested: testCase
  render () { // to be tested: use snapshot
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>  
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
