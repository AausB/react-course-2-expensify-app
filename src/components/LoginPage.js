import React from 'react';
import { connect } from 'react-redux';

import { startLogin } from '../actions/auth';

// destructure props object : { startLogin }
export const LoginPage = ({ startLogin }) => (
  <div>
    <button onClick={startLogin}>Login</button> 
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

// the connected component as default export
export default connect(undefined, mapDispatchToProps)(LoginPage);
