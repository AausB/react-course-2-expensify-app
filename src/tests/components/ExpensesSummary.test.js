import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary with one expense correctly', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={12345} />);

  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses correctly', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={10} expensesTotal={987654321} />);

  expect(wrapper).toMatchSnapshot();
});