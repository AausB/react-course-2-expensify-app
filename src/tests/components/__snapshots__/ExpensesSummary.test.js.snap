// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with multiple expenses correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        10
      </span>
       
      expenses
       totalling 
      <span>
        $9,876,543.21
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with one expense correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $123.45
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
